nmigen_boards/supercon19badge.py

```diff
@@ -0,0 +1,167 @@
+
+        # See note above for LED anode/cathode information.
+        # Short version is: these work as normal LEDs until you touch their cathodes.
+        *LEDResources(pins="E3 D3 C3 C4 C2 B1 B20 B19 A18 K20 K19",
+            attrs=Attrs(IO_TYPE="LVCMOS33")),
+        Resource("led_cathodes", 0, Pins("P19 L18 K18"), Attrs(IO_TYPE="LVCMOS33")),
+
+        UARTResource(0, rx="U2", tx="U1", attrs=Attrs(IO_TYPE="LVCMOS33")),
+
+        DirectUSBResource(0, d_p="F3", d_n="G3", pullup="E4", vbus_valid="F4",
+            attrs=Attrs(IO_TYPE="LVCMOS33")),
+
+        # Buttons, with semantic aliases.
+        *ButtonResources(pins="G2 F2 F1 C1 E1 D2 D1 E2",
+            attrs=Attrs(IO_TYPE="LVCMOS33", PULLMODE="UP")
+        ),
+        Resource("keypad", 0,
+            Subsignal("left",  Pins("G2", dir="i")),
+            Subsignal("right", Pins("F2", dir="i")),
+            Subsignal("up",    Pins("F1", dir="i")),
+            Subsignal("down",  Pins("C1", dir="i")),
+            Subsignal("start", Pins("E1", dir="i")),
+            Subsignal("select",Pins("D2", dir="i")),
+            Subsignal("a",     Pins("D1", dir="i")),
+            Subsignal("b",     Pins("E2", dir="i")),
+            Attrs(IO_TYPE="LVCMOS33", PULLMODE="UP")
+        ),
+
+        # Direct HDMI on the bottom of the board.
+        Resource("hdmi", 0,
+            Subsignal("clk", DiffPairsN("P20", "R20"), Attrs(IO_TYPE="TMDS_33")),
+            Subsignal("d", DiffPairs("N19 L20 L16", "N20 M20 L17"), Attrs(IO_TYPE="TMDS_33")),
+            Subsignal("hpd", PinsN("R18"), Attrs(IO_TYPE="LVCMOS33")),# Also called HDMI_HEAC_n
+            Subsignal("hdmi_heac_p", PinsN("T19"), Attrs(IO_TYPE="LVCMOS33")),
+            Attrs(DRIVE="4"),
+        ),
+
+        Resource("lcd", 0,
+            Subsignal("db",
+                Pins("J3 H1 K4 J1 K3 K2 L4 K1 L3 L2 M4 L1 M3 M1 N4 N2 N3 N1"),
+            ),
+            Subsignal("rd",    Pins("P2")),
+            Subsignal("wr",    Pins("P4")),
+            Subsignal("rs",    Pins("P1")),
+            Subsignal("cs",    Pins("P3")),
+            Subsignal("id",    Pins("J4")),
+            Subsignal("rst",   Pins("H2")),
+            Subsignal("fmark", Pins("G1")),
+            Subsignal("blen",  Pins("P5")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        Resource("spi_flash", 0, # clock needs to be accessed through USRMCLK
+            Subsignal("cs",   PinsN("R2")),
+            Subsignal("copi", Pins("W2")),
+            Subsignal("cipo", Pins("V2")),
+            Subsignal("wp",   Pins("Y2")),
+            Subsignal("hold", Pins("W1")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+        Resource("spi_flash_4x", 0, # clock needs to be accessed through USRMCLK
+            Subsignal("cs",   PinsN("R2")),
+            Subsignal("dq",   Pins("W2 V2 Y2 W1")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        # SPI-connected PSRAM.
+        Resource("spi_ram_4x", 0,
```

Comment:
I'd say call it `spi_psram_*` because the "PS" part is semantically important (you have to let go of CS every once in a while).

Reply:
Agreed. Changed.